network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,548 @@
+		return
+	}
+
+	if len(pod.Status.PodIP) > 0 {
+		podName := pod.GetName()
+		podCreationTime := pod.GetCreationTimestamp().Time
+		haveMaxTargets := false
+
+		c.targetPodsMap.lock.Lock()
+		_, targetUsed := c.targetPodsMap.mp[podName]
+		if !targetUsed {
+			c.targetPodsMap.mp[podName] = startTime
+			haveMaxTargets = len(c.targetPodsMap.mp) > c.config.expectedTargets
+		}
+		c.targetPodsMap.lock.Unlock()
+
+		// Process only up to the expected number of pods.
+		if haveMaxTargets {
+			c.stopInformerChan()
+			return
+		}
+
+		// Do the measurements only if it hasn't already been done for this pod.
+		if !targetUsed {
+			target := &targetSpec{
+				ip:        pod.Status.PodIP,
+				name:      podName,
+				namespace: pod.GetNamespace(),
+				startTime: startTime,
+			}
+			go c.recordFirstSuccessfulRequest(target, true)
+
+			podAssignedIPLatency := startTime.Sub(podCreationTime)
+			metrics.PodIPAddressAssignedLatency.Observe(podAssignedIPLatency.Seconds())
+			klog.Infof("Test client got pod %q with assigned IP %q, %v after pod creation", podName, pod.Status.PodIP, podAssignedIPLatency)
+		}
+	}
+}
+
+// recordFirstSuccessfulRequest sends curl requests continuously to the provided
+// target IP and target port from the test client config, at 1 second intervals,
+// until the first successful response. It records the latency between the
+// provided start time in the targetSpec, and the time of the first successful
+// response.
+func (c *TestClient) recordFirstSuccessfulRequest(target *targetSpec, forPodCreation bool) {
+	if target == nil || len(target.ip) == 0 {
+		klog.Infof("Skipping a target for policy enforcement latency (podCreation=%t), because the target does not have an IP address assigned. Target: %v", forPodCreation, target)
+		return
+	}
+
+	request := fmt.Sprintf("curl --connect-timeout 1 %s:%d", target.ip, c.config.targetPort)
+	command := []string{"sh", "-c", "-x", request}
+	klog.Infof("Sending requests %q", request)
+
+	for {
+		select {
+		case <-c.MainStopChan:
+			return
+		default:
+		}
+
+		_, err := utils.RunCommand(command)
+		if err == nil {
+			if forPodCreation {
+				c.reportReachedTimeForPodCreation(target, time.Now())
+			} else {
+				c.reportReachedTimeForPolicyCreation(target, time.Now())
+			}
+			return
+		}
+	}
+}
+
+// reportReachedTimeForPodCreation records network policy enforcement latency
+// metric for pod creation.
+func (c *TestClient) reportReachedTimeForPodCreation(target *targetSpec, reachedTime time.Time) {
+	latency := reachedTime.Sub(target.startTime)
+	metrics.PolicyEnforceLatencyPodCreation.Observe(latency.Seconds())
+	klog.Infof("Pod %q in namespace %q reached %v after pod IP was assigned", target.name, target.namespace, latency)
+}
+
+// reportReachedTimeForPolicyCreation records network policy enforcement latency
+// metric for network policy creation.
+func (c *TestClient) reportReachedTimeForPolicyCreation(target *targetSpec, reachedTime time.Time) {
```

Comment:
Why not to return errors?

Reply:
I don't see the benefit of returning an error here. Logging a warning in case it fails is enough.
It's used in `recordFirstSuccessfulRequest()`, which is called many times as a separate go routine and won't impact the main execution.